src/cpp/src/llm_pipeline.cpp

```diff
@@ -101,12 +119,17 @@ class StatefulLLMPipeline final : public LLMPipelineImplBase {
                 m_history.push_back({{"role", "user"}, {"content", prompt}});
                 constexpr bool add_generation_prompt = true;
                 auto new_templated_chat_history  = m_tokenizer.apply_chat_template(m_history, add_generation_prompt);
-                
-                prompt = new_templated_chat_history.substr(m_templated_chat_history.size());
+                auto new_chat_tokens = m_tokenizer.encode(new_templated_chat_history);
+                if (m_is_cache_empty) {
+                    encoded_input = new_chat_tokens;
+                } else {
+                    auto prev_chat_tokens = m_tokenizer.encode(m_templated_chat_history);
+                    encoded_input = subtract_chat_tokenized_inputs(new_chat_tokens, prev_chat_tokens);
```

Comment:
Why is the history stored as `std::string` instead of encoded tensors?
@pavel-esir, maybe you can explain?

Reply:
I wrote this code originally. 
std::string is used because after tokenization on next step we will get the same tokens we have in kv-cache. 
Encoded tensors don't contain <eos> token at the end by default. And kv-cache contains it. So we have add it manually.

Reply:
Please add the explanation to the code comments.

Reply:
done